examples/animatediff/ad/modules/diffusionmodules/motion_module.py

```diff
@@ -0,0 +1,523 @@
+    module.bias.set_data(bias_weight)
+    return module
+
+
+def get_motion_module(in_channels, motion_module_type: str, motion_module_kwargs: dict):
+    if motion_module_type == "Vanilla":
+        return VanillaTemporalModule(
+            in_channels=in_channels,
+            **motion_module_kwargs,
+        )
+    else:
+        raise ValueError
+
+
+class VanillaTemporalModule(nn.Cell):
+    def __init__(
+        self,
+        in_channels,
+        num_attention_heads=8,
+        num_transformer_block=2,
+        attention_block_types=("Temporal_Self", "Temporal_Self"),
+        cross_frame_attention_mode=None,
+        temporal_position_encoding=False,
+        temporal_position_encoding_max_len=24,
+        temporal_attention_dim_div=1,
+        zero_initialize=True,
+    ):
+        super().__init__()
+
+        # print("D---: zero_initialize in MM", zero_initialize)
+        self.temporal_transformer = TemporalTransformer3DModel(
+            in_channels=in_channels,
+            num_attention_heads=num_attention_heads,
+            attention_head_dim=in_channels // num_attention_heads // temporal_attention_dim_div,
+            num_layers=num_transformer_block,
+            attention_block_types=attention_block_types,
+            cross_frame_attention_mode=cross_frame_attention_mode,
+            temporal_position_encoding=temporal_position_encoding,
+            temporal_position_encoding_max_len=temporal_position_encoding_max_len,
+        )
+
+        if zero_initialize:
+            self.temporal_transformer.proj_out = zero_module(self.temporal_transformer.proj_out)
+
+    def construct(
+        self, input_tensor, temb, encoder_hidden_states, attention_mask=None, anchor_frame_idx=None, video_length=None
+    ):
+        """
+        Args:
+            hidden_states: x, (b*f c h w)
+            temb: time embedding, not used
+            encoder_hidden_states: prompt embedding, not used
+        Returns:
+            (b*f c h w)
+        """
+        dtype = input_tensor.dtype
+
+        hidden_states = input_tensor
+        hidden_states = self.temporal_transformer(
+            hidden_states, encoder_hidden_states, attention_mask, video_length=video_length
+        )
+
+        output = hidden_states
+
+        output = output.to(dtype)  # TODO: need debug for fp16 or fp32
+        return output
+
+
+class FeedForward(nn.Cell):
+    def __init__(
+        self,
+        dim: int,
+        dim_out=None,
+        mult: int = 4,
+        dropout: float = 0.0,
+        activation_fn: str = "geglu",
+        dtype=ms.float32,
+    ):
+        super().__init__()
+        inner_dim = int(dim * mult)
+        dim_out = dim_out if dim_out is not None else dim
+
+        if activation_fn == "geglu":
+            act_fn = GEGLU(dim, inner_dim, approximate=False, dtype=dtype)
+        else:
+            raise NotImplementedError
+
+        self.net = nn.CellList([])
+        # project in
+        self.net.append(act_fn)
```

Comment:
Maybe the input to this `act_fn` needs to be cast to fp32.

Reply:
geglu should be fine. the tested mre error < 1e-4